 <img align="left" width="75" height="75"  src="https://upload.wikimedia.org/wikipedia/en/c/c8/University_of_the_Punjab_logo.png"> 

<h1 align="center">Department of Data Science</h1>
<h1 align="center">Course: Tools and Techniques for Data Science</h1>

---
<h3><div align="right">Instructor: Muhammad Arif Butt, Ph.D.</div></h3>    

<h1 align="center">Lecture 6.25 (A Comprehensive Recap of Probability for Bayes' Theorem)</h1>

<a href="https://colab.research.google.com/github/arifpucit/data-science/blob/master/Section-4-Mathematics-for-Data-Science/Lec-4.1(Descriptive-Statistics).ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# <img align="center" width="900" src="images/ml-topimg1.png"  >

## Learning agenda of this notebook

**A Comprehensive Recap of Probability and Bayes' Theorem**

- Overview of Probability and Statistics
    - Classical vs Emperical Probability
    - Some Important Terms related to Probability
- Marginal/Simple/Unconditional Probability (with examples)
- Types of Events
    - Mutually Exclusive vs Not Mutually Exclusive (with examples)
    - Independent and Dependent Events (with examples)
  
- Joint Probability
    - Reducing the Sample Space
    - Use of Venn Diagrams and Contigency Tables
    - Examples
- Conditional Probability
    - Intuition
    - Examples

# 1. Overview of Probability and Statistics
>- <font color=green size=3> Descriptive Statistics:</font> https://www.youtube.com/watch?v=Q6EZ-KGm3UE&list=PL7B2bn3G_wfDTEU20jwGcUKlVqdAyQMax&index=1
>- <font color=green size=3>Inferential Statistics:</font> https://www.youtube.com/watch?v=y4t9ELO3Nxc&list=PL7B2bn3G_wfDTEU20jwGcUKlVqdAyQMax&index=2

- **Statistics** is the area of applied mathematics that deals with data collection & organization, data analysis, data interpretation and data representation.
    - `Descriptive Statistics` is used to describe, summarize, and present different datasets through numerical calculations and graphs.
    - `Inferential Statistics` is used to make prediction and inferences for an entire population based on sample data from that population

<img align="centre" width="1000"  src="images/di.png"  >

## a. Classical vs Emperical Probability?

<img align="center" width="400" height="200"  src="images/prob.gif"  > 

- Probability theory is a branch of mathematics concerned with the analysis of random phenomena, and defines the `likelihood of occurrence of an event`. 

- The value of the probability of an event to happen can lie between 0 and 1 because the favorable number of outcomes can never cross the total number of outcomes. 
- When the probability of something approaches 1, then it means it is very likely, and when the probability of something approaches 0, then it means that it is very unlikely.
- To understand Probability, we normally start to predict the outcomes for the `tossing of coins`, `rolling of dice`, or `drawing a card from a pack of playing cards`.


<img align="center" width="400"  src="images/prob1.png"  >


- **Classical/Theoratical Probability:**
    - It is a probability measure that is used when you `DO NOT have past data` and is based on theoratical assumptions
<font face = "Arial" color=magenta size=3>$$ P(\text{event}) = \frac{\text{Number of favourable outcomes of event}}{\text{Total outcomes in sample space}} $$</font>

    - Consider a scenario of tossing a coin hundred times (without performing the experiment). The theoratical probability of getting a head is  : $ P(H) = \large \frac{50}{100} \small = 0.5 $



- **Emperical/Experimental Probability:**
    - It is a probability measure that is `based on observed/past data`. It donot use sample space rather use frequency distributions 
<font face = "Arial" color=blue size=3>$$ P(\text{event}) = \frac{\text{Frequency of event}}{\text{Sum of the frequencies}} $$</font>
    - Suppose we toss a coin hundred times and get heads 60 times and tails 40 times. The emperical probability of getting a head is: $ P(H) = \large \frac{60}{100} \small = 0.6 $

    >- When we `increase the number of trials/tosses` the result get close and close to 0.5. This is called the `Law of Large Numbers`, which states that the more experiments we run, the closer we will tend to get to the Classical probability. 

## b. Important Terms Related to Probability

- **Random Experiment:** An experiment is called random experiment if it satisfies following two conditions:
    - It has more than one possible outcome.
    - It is not possible to predict the outcome in advance.
- **Trial:** It refers to a single execution of a random experiment. Each trial produces an outcome.
- **Sample Space:** It is the `set of all possible outcomes` that can occur. Generally one random experiment will have one set of sample space. Sample space for tossing a coin is {H, T}, sample space for rolling a die is {1,2,3,4,5,6}, sample space for tossing two coins is {HH,HT,TH,TT}.
- **Outcome:** It is the `result of a single trial of a random experiment`.
- **Event:**  Event is a `subset of sample space`  and can include a single outcome or multiple outcomes. One random experiment can have multiple events. 

### Tossing a Coin Twice:

 <img align="center" width="800"  src="images/prob3a.png"  >

## Rolling two Dice:

 <img align="center" width="800"  src="images/prob2b.png"  >

## Titanic Dataset:

<img align="center" width="900"  src="images/titanicimg.png"  >
<br><br>

- **Random Experiment:** Randomly selecting a passenger and finding his/her pclass
- **Trial:** Randomly selecting a passenger and finding his/her pclass (once)
- **Sample Space:** {1, 2, 3}
- **Outcome:** {1}
- **Event:** 
    - Getting a passenger belonging to pclass 2: {2}
    - Getting a passenger NOT belonging to pclass 3: {1, 2}

# 2. Marginal/Simple/Unconditional Probability
<h3 align="center"><div class="alert alert-success" style="margin: 20px">Marginal Probability is the probability of a single event occurring, independent of the outcome of other events.</h3>

<img align="right" width="250"  src="images/prob7.png"  > 

**Example 1:** Consider a scenario of tossing a fair coin three times.

- Event A: Getting exactly two heads: $ \hspace{2cm} P(A) \hspace{0.3cm}=\hspace{0.3cm} \large \frac{3}{8} \small \hspace{0.3cm}=\hspace{0.3cm} 0.375 $<br><br>

- Event B: Getting three consecutive heads: $ \hspace{1cm} P(B) \hspace{0.3cm}=\hspace{0.3cm} \large \frac{1}{8} \small \hspace{0.3cm}=\hspace{0.3cm} 0.125 $<br><br>

- Event C: Getting at least two heads: $\hspace{2cm} P(C) \hspace{0.3cm}=\hspace{0.3cm} \large \frac{4}{8} \small \hspace{0.3cm}=\hspace{0.3cm} 0.5 $

**Example 2:** An urn contains 2 red balls, 3 green balls, and 5 blue balls. A single ball is selected at random `with replacement`
<img align="right" width="200"  src="images/prob6.png"  > 

- Event A: Getting a red ball: $ \hspace{2cm} P(A) \hspace{0.3cm}=\hspace{0.3cm} \large \frac{2}{10} \small \hspace{0.3cm}=\hspace{0.3cm} 0.2 $<br><br>

- Event B: Getting a green ball: $ \hspace{1.7cm} P(B) \hspace{0.3cm}=\hspace{0.3cm} \large \frac{3}{10} \small \hspace{0.3cm}=\hspace{0.3cm} 0.3 $<br><br>

- Event C: Getting a blue ball: $ \hspace{2cm} P(C) \hspace{0.3cm}=\hspace{0.3cm} \large \frac{5}{10} \small \hspace{0.3cm}=\hspace{0.3cm} 0.5 $

**Example 3:** Consider drawing a single card from a standard deck of 52 playing cards (Clubs, Spades, Hearts, Diamonds). In this case, the number of possible outcomes in the sample space $\Omega$ is 52. 

<img align="right" width="400"  src="images/ssplayingcards.png"  > 

- Event A: Getting a `spades` card: $ \hspace{2cm} P(A) \hspace{0.3cm}=\hspace{0.3cm} \large \frac{13}{52} \small \hspace{0.3cm}=\hspace{0.25cm} 0.25 $<br><br>

- Event B: Getting a `ace` card: $ \hspace{2.7cm} P(B) \hspace{0.3cm}=\hspace{0.3cm} \large \frac{4}{52} \small \hspace{0.3cm}\approx\hspace{0.25cm} 0.0769 $<br><br>


- Event C: Getting a `Queen of Hearts` card: $ \hspace{0.1cm} P(C) \hspace{0.3cm}=\hspace{0.3cm} \large \frac{1}{52} \small \hspace{0.3cm}\approx\hspace{0.25cm} 0.019 $<br><br>

- Event D: Getting a card: $ \hspace{4cm} P(D) \hspace{0.3cm}=\hspace{0.3cm} \large \frac{52}{52} \small \hspace{0.3cm}=\hspace{0.25cm} 1.0 $<br><br>

- Event E: Getting an `Apple`: $ \hspace{3cm} P(E) \hspace{0.3cm}=\hspace{0.3cm} \large \frac{0}{52} \small \hspace{0.3cm}\approx\hspace{0.25cm} 0.0 $

**Example 4:** Consider a scenario of rolling two fair 6-sided dice. 

<img align="center" width="800" src="images/ss-twodice.png"  > 


- Event A: Getting a sum of 3: $ \hspace{2cm} P(A) \hspace{0.3cm}=\hspace{0.3cm} \large \frac{2}{36} \small \hspace{0.3cm}\approx\hspace{0.25cm} 0.05555 $<br><br>

- Event B: Getting a sum greater than 9: $ \hspace{0.5cm} P(B) \hspace{0.3cm}=\hspace{0.3cm} \large \frac{6}{36} \small \hspace{0.3cm}\approx\hspace{0.25cm} 0.16666 $<br><br>

- Event C: Getting a odd sum: $ \hspace{2cm} P(C) \hspace{0.3cm}=\hspace{0.3cm} \large \frac{18}{36} \small \hspace{0.3cm}=\hspace{0.25cm} 0.5 $

# 3. The Concept of Random Variables and Probability Distributions

## a. Random Variable


<h3 align="center"><div class="alert alert-success" style="margin: 20px">A random variable is a mathematical function that assigns a numerical value to every possible outcome of a random experiment</h3>


<img align="center" width="400" src="images/random-variable-definition.png"  > 

- Probability Distribution for the random experiment of tossing a coin is as follows:

$$P(X = 0) = 1 / 2$$ 
$$P(X = 1) = 1 / 2$$


## b. Types of Random Variables
    
<img align="center" width="600" src="images/random-variable-definition2.png"  > 
    
    

**Discrete Random Variable:** 
- A `discrete random variable` is a variable which can take on discrete values as outcome, which can be counted not measured, e.g.,  $0, 1, 2, 3, 4, 5, \cdots$ 
- Examples:
    - The number of times a coin lands on tails after being flipped 20 times.
    - The number of times a dice lands on the number 4 after being rolled 100 times.
    - Number of items sold at a store on a certain day. Using historical sales data, a store could create a probability distribution that shows how likely it is that they sell a certain number of items in a day.
    - Number of customers that enter a shop on a given day. Using historical data, a shop could create a probability distribution that shows how likely it is that a certain number of customers enter the store.
    - Number of defective products produced per batch by a certain manufacturing plant. Using historical data on defective products, a plant could create a probability distribution that shows how likely it is that a certain number of products will be defective in a given batch.
    - Number of traffic accidents that occur in a specific city on a given day. Using historical data, a police department could create a probability distribution that shows how likely it is that a certain number of accidents occur on a given day.





**Continouous Random Variable:** 
- A `continuous random variable` is a variable which can take on an infinite number of possible values as outcome, which can be measured not counted, e.g.,  $0,0.1,0.2,3.5,4.2,4.3,\cdots$
- Examples:
    - Weight of a certain animal like a dog is a continuous random variable because it can take on an infinite number of values (e.g., 30.333 pounds, 50.340999 pounds, 60.5 pounds). In this case, we could collect data on the weight of dogs and create a probability distribution that tells us the probability that a randomly selected dog weighs between two different amounts.
    - Height of a certain species of plant is a continuous random variable because it can take on an infinite number of values (e.g., 6.5555 inches, 8.95 inches, 12.32426 inches). In this case, we could collect data on the height of this species of plant and create a probability distribution that tells us the probability that a randomly selected plant has a height between two different values.
    - Distance traveled by a certain wolf during migration season is a continuous random variable because it can take on an infinite number of values (e.g., a wolf may travel 40.335 miles, 80.5322 miles, 105.59 miles). In this scenario, we could collect data on the distance traveled by wolves and create a probability distribution that tells us the probability that a randomly selected wolf will travel within a certain distance interval.

## c. Probability Distribution
<center><font face = "Garamond" color=red size=4.5><div class="alert alert-success" style="margin: 10px">A probability distribution tells us the probability of all the possible values of a random variable, and sums up to 1</font></center>

<img align="center" width="900" src="images/prob-distr-types1.png"  > 

    


### Example of Probability Distribution for Discrete Random Variable

<img align="center" width="700" src="images/ss-twodice.png"  > 

<img align="center" width="600" src="images/prob_distr2.png"  > 

<br>

- The probability that the sum is exactly 2: $P(X=2)=\large \frac{1}{36} \small =0.03$
- The probability that the sum is exactly 3: $P(X=3)=\large \frac{2}{36} \small =0.06$
- The probability that the sum is exactly 4: $P(X=4)=\large \frac{3}{36} \small =0.08$
- The probability that the sum is exactly 5: $P(X=5)=\large \frac{4}{36} \small =0.11$

# 4. Joint Probability
<h3 align="center"><div class="alert alert-success" style="margin: 20px">Joint Probability is the probability of two events occurring simultaneously</h3>

<font face = "Arial" color=green size=4> $ \hspace{10cm}    P(X=x, Y=y) \hspace{0.5cm} $</font>

## a. Mutually Exclusive vs Not Mutually Exclusive Events
### (i) Mutually Exclusive Events: 
- Mutually Exclusive events are events that `CANNOT` occur at the same time.
- **Examples:**
    - If a die is rolled the two events getting a `3` and getting a `5` are mutually exclusive. 
    - If a card is selected at random from a deck of 52 playing cards, the probability that the card is an `Ace` or a `King` are mutually exclusive. 
      
<font face = "Arial" color=magenta size=4>$ \hspace{3.5cm}$ Addition rule: $ \hspace{2cm}    P(A \cup B) \hspace{0.5cm} = \hspace{0.5cm}  P(A) + P(B)$</font>

### (ii) Not Mutually Exclusive Events: 
- Not Mutually Exclusive events are events that `CAN` occur at the same time.
- **Example:**
    - If a card is selected at random from a deck of 52 playing cards, the probability that the card is a `Queen` or a card of `Hearts` are not mutually exclusive.

<font face = "Arial" color=magenta size=4>$ \hspace{3.5cm}$Addition rule: $ \hspace{2cm}    P(A \cup B) \hspace{0.5cm} = \hspace{0.5cm}  P(A) + P(B) - P(A \cap B)$</font>

<br><br>
**Note:** `Addition rule is used to find the probability of at least one of two events occurring`

## b. Independent vs Dependent Events
### (i) Independent Events: 
- Two events are said to be `independent`, when the occurrence of one event `DOES NOT` affect the probability of occurrence of other event.

- **Examples:**
    - If a coin is tossed and then a die is rolled, the outcome of the coin in no way affects or changes the probability of the outcome of the die. 
    - Selecting a card from a deck, replacing it, and then selecting a second card from a deck are two independent events.
    - Drawing a ball from an urn, replacing it, and then drawing a second ball.    
    
<font face = "Arial" color=blue size=4>$ \hspace{3cm}$Multiplication Rule: $ \hspace{1cm}    P(A \cap B) \hspace{0.5cm} = \hspace{0.5cm}  P(A) \times P(B)$</font><br><br>

### (ii) Dependent Events: 
- Two events are said to be `dependent`, if they happen one after another, and the occurrence of one event `DOES` affect the probability of occurrence of other event.
 <br><br>
- **Examples:**
    - Drawing a ball from an urn, not replacing it, and then drawing a second ball.
    - Drawing two cards without replacement from a deck of cards.
     
    
<font face = "Arial" color=blue size=4>$ \hspace{3cm}$Multiplication Rule: $ \hspace{1cm}   P(A \cap B) \hspace{0.5cm} = \hspace{0.5cm}  P(A) \times P(B \mid A)$</font><br><br>


<br><br>
**Note:** `Multiplication rule is used to find the probability of two events occurring in sequence`


**Example 1:** An urn contains 2 red balls, 3 green balls, and 5 blue balls. 
<img align="right" width="250"  src="images/prob6.png"  > 


- Find the `joint probability` if two balls are drawn one after another `with replacement` (independent events):
    - Selecting a blue ball and then a red ball:   $\hspace{1cm} P(B \cap R) = P(B) \times P(R) =\large\frac{5}{10} \small \times \large\frac{2}{10} \small= \large\frac{1}{10} $
    - Selecting a green ball and then a blue ball: $\hspace{0.5cm} P(G \cap B) = P(G) \times P(B) =\large\frac{3}{10} \small \times \large\frac{5}{10} \small= \large\frac{3}{20} $
    - Selecting a blue ball and then another blue ball: $ P(B \cap B) = P(B) \times P(B) =\large\frac{5}{10} \small \times \large\frac{5}{10} \small= \large\frac{1}{4} $
<br><br> <br>  
- Find the `joint probability` if two balls are drawn one after another `without replacement` (dependent events):
    - Selecting a blue ball and then a red ball:   $\hspace{2cm} P(B \cap R) = P(B) \times P(R\mid B) = \large\frac{5}{10} \small \times \large\frac{2}{9} \small= \large\frac{1}{9} $
    - Selecting a green ball and then a blue ball: $\hspace{1.5cm} P(G \cap B) = P(G) \times P(B\mid G) =\large\frac{3}{10} \small \times \large\frac{5}{9} \small= \large\frac{3}{18} $
    - Selecting a blue ball and then another blue ball:   $\hspace{0.5cm} P(B \cap B) = P(B) \times P(B\mid B) = \large\frac{5}{10} \small \times \large\frac{4}{9} \small= \large\frac{2}{9} $

**Example 2 (Dependent Events):** Two cards are drawn `without` replacement from a deck of 52 cards. Find the probability that both are queens.
- Event A: First card is Queen
- Event B: Second card is Queen

$ \hspace{2cm}  P(A \cap B) \hspace{0.5cm} = \hspace{0.5cm} P(A) \times P(B \mid A) \hspace{0.5cm} = \hspace{0.5cm} \large \frac{4}{52} \small\times \large\frac{3}{51} \small \hspace{0.5cm} = \hspace{0.5cm} \large\frac{12}{2652}\small \hspace{0.5cm} = \hspace{0.5cm} 0.0045$

**Example 3 (Dependent Events):** A box contains 24 toasters, 3 of which are defective. If two toasters are selected and tested, find the probability that both are defective (assume toasters are `not replaced`).

- Event A: First toaster is defective
- Event B: Second toaster is defective

$ \hspace{2cm}  P(A \cap B) \hspace{0.5cm} = \hspace{0.5cm} P(A) \times P(B \mid A) \hspace{0.5cm} = \hspace{0.5cm} \large \frac{3}{24} \small\times \large\frac{2}{23} \small \hspace{0.5cm} = \hspace{0.5cm} \large\frac{1}{92}\small \hspace{0.5cm} = \hspace{0.5cm} 0.01087$

**Example 4 (Independent Events):** A coin is tossed three times and a die is rolled twice. Find the probability of getting exactly two heads on the coin and a sum of 5 on the dice.

<img align="center" width="800"  src="images/prob4.png"  > 


<br><br>


- Event A: Getting exactly two heads on the coin: $\hspace{1cm} P(A) \hspace{0.5cm}=\hspace{0.5cm} \large\frac{3}{8} $<br><br>
- Event B: Getting a sum of 5 on the dice: $\hspace{2.4cm} P(B) \hspace{0.5cm}=\hspace{0.5cm} \large\frac{4}{36} $<br><br>

- Joint Probability of the two events:$\hspace{3cm} P(A \cap B) \hspace{0.5cm}=\hspace{0.5cm} P(A) \times P(B) \hspace{0.5cm}=\hspace{0.5cm} \large\frac{3}{8} \small\times \large\frac{4}{36} \small \hspace{0.5cm}\approx\hspace{0.5cm} 0.04166$

**Example 5:** Consider a random experiment of rolling two fair six-sided dice. Also consider the random variable is the sum of the two numbers that appeared on the two dice.
<img align="center" width="800" src="images/ss-twodice.png"  > 

- Event A: Getting a sum of greater than or equal to 5, i.e., {5,6,7,8,9,10,11,12}
- Event B: Getting an even number, i.e., {2,4,6,8,10,12}
- `Calculate the probability of the two events occurring in sequence (one after another):` $\large P(A\cap B)$    
- `Calculate the probability of at least one of the two events occurring:` $\large P(A\cup B)$ 


<img align="center" width="800" src="images/contingency-table2.png"  > 

**Calculating Probabilities using Venn-Diagram and Contingency Table:**<br>
$\hspace{2cm} P(A) = \large \frac{30}{36}\small=0.8333$,
$\hspace{1cm} P(B) = \large \frac{18}{36}\small=0.5$<br><br>

**Calculating Probabilities using Addition and Multiplication Rules:**<br>
$\hspace{2cm} P(A \cup B) \hspace{0.5cm}=\hspace{0.5cm} P(A) +P(B)-P(A\cap B)\hspace{0.5cm}=\hspace{0.5cm}\large \frac{30}{36}\small+\large\frac{18}{36}\small-\large\frac{14}{36}\small\hspace{0.7cm}=\hspace{0.5cm}\large\frac{34}{36}\small\hspace{0.5cm}=\hspace{0.5cm}0.944$<br>

$\hspace{2cm} P(A \cap B) \hspace{0.5cm}=\hspace{0.5cm} P(A) \times P(B\mid A)\hspace{2cm}=\hspace{0.5cm}\large \frac{30}{36}\small\times\large\frac{14}{30}\small\hspace{2cm}=\hspace{0.5cm}\large\frac{7}{18}\small\hspace{0.5cm}=\hspace{0.5cm}0.389$<br>


<img align="center" width="600" src="images/contingency-table13.png"  > 

**Example 6:** 
- A class contains 100 students; 70 of them like mathematics, 60 like physics, and 40 like both. 
    - Event A: Student like mathematics
    - Event B: Student like physics
- `Calculate the probability of the two events occurring in sequence (one after another):` $\large P(A\cap B)$    
- `Calculate the probability of at least one of the two events occurring:` $\large P(A\cup B)$ 

<img align="center" width="800" src="images/contingency-table11.png"  > 

- If a student is chosen at random, find the probability that he/she 
    - like mathematics $= P(A) = \large \frac{70}{100}$ 
    - like physics $= P(B) = \large \frac{60}{100}$
    - like both mathematics and physics $= P(A \cap B)= \large \frac{40}{100}$
    - like mathematics or physics $= P(A \cup B)= P(A) +P(B)-P(A\cap B)=\large \frac{70}{100}\small+\large \frac{60}{100}\small-\large \frac{40}{100}\small=\large \frac{90}{100}$
    - like mathematics but not physics $= \large \frac{30}{100}$
    - like physics but not mathematics $= \large \frac{20}{100}$
    - like neither physics nor mathematics $= \large \frac{10}{100}$
    
    
<img align="center" width="600" src="images/contingency-table12.png"  > 

**Example 7:** Suppose `Netflix` took a survey of their subscribers to determine people favourite movies. Load the following toy dataset, create contingency table using the Panda's `crosstab` method, and calculate different joint and marginal probabilities.
>- Let X be a random variable associated with the `movie` and can take three different values
>- Let Y be a random variable associated with the `gender` and can take two values

In [ ]:
import numpy as np
import pandas as pd

In [ ]:
df = pd.read_csv('datasets/sample1.csv')
df

>- From above dataset, construct contingency table by keeping `movie` along horizontal rows, while `gender` along vertical columns

In [ ]:
pd.crosstab(index=df['movie'], columns=df['gender'])

In [ ]:
pd.crosstab(index=df['movie'], columns=df['gender'], margins='all')

In [ ]:
# normalize = {'all', 'index', columns'}
pd.crosstab(index=df['movie'], columns=df['gender'], margins=True, normalize='all')

**Example 8:** Load the Titanic dataset, create contingency table using the Panda's `crosstab` method (for the survived and passenger class columns), and calculate different joint and marginal probabilities.
>- Let X be a random variable associated with the `pclass` of a Titanic passenger and can take the values {1,2,3}
>- Let Y be a random variable associated with the `survived` status of a Titanic passenger and can take the values {0, 1} 

In [ ]:
import numpy as np
import pandas as pd
df = pd.read_csv('datasets/titanic2.csv')
df

In [ ]:
df1 = pd.concat([df['pclass'], df['sex'], df['survived']], axis=1)
df1

In [ ]:
pd.crosstab(index=df['pclass'], columns=df['survived'], margins=True)

In [ ]:
pd.crosstab(index=df['survived'], columns=df['pclass'], margins=True)

>-  The cell value of 80 tells us that there are 80 passengers who belong to pclass=1 and do not survived.
>-  The cell value of 136 tells us that there are 136 passengers who belong to pclass=1 and survived.
>- and so on....
>- Now this tells us the count, and to calculate the probability we simply has to divide this count with the total number of passenger onboard and i.e., 891 and you get 0.08978. which is the probability of a passenger belonging to class 1 and who do not survived. This is  what joint probability is, where you take more than one event and you calculate the probability of happening of both the events

In [ ]:
pd.crosstab(index=df['survived'], columns=df['pclass'], margins=True, normalize='all')

**If a passenger is chosen at random, find the `joint probability` that he/she:**<br><br>
    - belongs to `pclass 1` and `not survived`: $P(\text{pclass}=1, \text{survived}=0)=0.089787$ <br>
    - belongs to `pclass 2` and `not survived`: $P(\text{pclass}=2, \text{survived}=0)=0.108866$ <br>
    - belongs to `pclass 3` and `not survived`: $P(\text{pclass}=3, \text{survived}=0)=0.417508$  <br>
    - belongs to `pclass 1` and `survived`: $P(\text{pclass}=1, \text{survived}=1)\hspace{0.9cm}=0.152637$  <br>
    - belongs to `pclass 2` and `survived`: $P(\text{pclass}=2, \text{survived}=1)\hspace{0.9cm}=0.097643$  <br>
    - belongs to `pclass 3` and `survived`: $P(\text{pclass}=3, \text{survived}=1)\hspace{0.9cm}=0.133558$

>- Each of above six individual numbers (probabilities) is called joint probability, and all the six probabilities together is known as `Joint Probability Distribution`, and note the sum of these six probability values sums upto one.

**If a passenger is chosen at random, find the `marginal probability` that he/she:**<br><br>
- died, irrespective of pclass he/she is travelling: $\hspace{1.9cm}P(\text{survived=0})=0.616162$<br>
- survived, irrespective of pclass he/she is travelling: $\hspace{1.3cm}P(\text{survived=1})=0.383838$<br>
- travels in pclass 1, irrespective he/she died or survived: $\hspace{.5cm}P(\text{pclass=1})=0.242424$<br>
- travels in pclass 2, irrespective he/she died or survived: $\hspace{.5cm}P(\text{pclass=2})=0.206510$<br>
- travels in pclass 3, irrespective he/she died or survived: $\hspace{.5cm}P(\text{pclass=3})=0.551066$<br>

>-  0.616162 and 0.383838 is called the `Marginal Probability Distribution` of random variable `survived` and note the sum of these two probability values is 1.
>-  0.242424,  0.206510 and 0.551066 is called the `Marginal Probability Distribution` of random variable `pclass`  and the sum of these three probability values is 1.

# 6. Conditional Probability
<h3 align="center"><div class="alert alert-success" style="margin: 20px">Conditional Probability is the probability of an event occurring given that another event has already occurred.</h3>
    
<img align="center" width="600"  src="images/cloudy-example.png"  > 

> In `joint probability` the two events A and B were occurring together, while in `conditional probability` we normally say that  event B has already occurred and  we see its impact on the probability of occurring of event A

<img align="right" width="300"  src="images/prob7.png"  > 

**Example 1:** Three unbiased coins are tossed. What is the conditional probability that at least two coins show heads, given that at least one coin shows heads?

- Event-A: Atleast two heads {HHH, HHT, HTH, THH}
- Event-B: At least one heads (Event has already happened): {HHH, HHT, HTH, THH, HTT, THT, TTH}

<br><br>

**Calculate probability of event A given that event B has already occurred:**

$$ P(A \mid B) =  \frac{4}{7} = 0.5714 $$

**Calculate probability of event B given that event A has already occurred:**

$$ P(B \mid A) =  \frac{4}{4} = 1.0 $$

**Example 2:** Two fair six-sided dice are rolled. What is the conditional probability that the die-1 equals 2, given that the sum of numbers on two dice is less than or equal to 5?

- Event-A: Number on first die is 2: $\{(2,1), (2,2), (2,3), (2,4), (2,5), (2,6)\}$
- Event-B: Sum of numbers on two dice is less than or equal to 5: $\{(1,1),(1,2),(1,3),(1,4),(2,1),(2,2),(2,3),(3,1),(3,2),(4,1)\}$

<img align="center" width="800" src="images/ss-twodice.png"  > 


**Calculate probability of event A given that event B has already occurred:**

$$ P(A \mid B) = \frac{3}{10} = 0.3 $$

**Calculate probability of event B given that event A has already occurred:**

$$ P(B \mid A) = \frac{3}{6} = 0.5 $$


**Calculate Conditional Probability using Contingency Table:**
<img align="center" width="500" src="images/cond-prob-tbl1.png"  > 

**Example 3:** Two fair six-sided dice are rolled. What is the conditional probability that the sum of the numbers rolled is 8, given that the sum is an even number.

- Event-A: Sum is 8 ${(2,6),(3,5),(4,4),(5,3),(6,2)}$
- Event-B: Sum is an even number (Eighteen possible outcomes)


<img align="center" width="800" src="images/ss-twodice.png"  > 

**Calculating $ P(A \mid B)$ using the reduced Sample Space:**
- Since event B has already occurred, so the new reduced sample space has 18 total outcomes instead of 36
$$ P(A \mid B) = \frac{5}{18} = 0.2778 $$

**Calculating $ P(B \mid A)$ using the reduced Sample Space:**
- Since event A has already occurred, so the new reduced sample space has 5 total outcomes instead of 36
$$ P(B \mid A) = \frac{5}{5} = 1.0 $$




**Calculate Conditional Probability using Contingency Table:**
<img align="center" width="500" src="images/cond-prob-tbl2.png"  >

**Example 4:** Consider the given contingency table containing data of 200 male and female birds in the zoo having brown and blue eyes.

<img align="center" width="600" height="400"  src="images/cond-prob-ex.png"  >

**Compute different conditional probabilities of a randomly selected bird being male/female given that it has brown/blue eyes:**<br><br>

- $\hspace{.3cm}P(\text{X=male} \mid \text{Y=brown}) \hspace{0.9cm}=\hspace{0.5cm}\large \frac{P(\text{X=male} \hspace{0.2cm}\cap\hspace{0.2cm} \text{Y=brown})}{P(\text{Y=brown})}       \hspace{0.5cm} \small= \hspace{0.5cm} \large \frac{70}{170}\hspace{0.2cm} \small \approx \hspace{0.2cm} 0.4117$<br>


- $\hspace{.3cm}P(\text{X=female} \mid \text{Y=brown}) \hspace{0.5cm}=\hspace{0.5cm}\large \frac{P(\text{X=female} \hspace{0.2cm}\cap\hspace{0.2cm} \text{Y=brown})}{P(\text{Y=brown})}       \hspace{0.3cm} \small= \hspace{0.5cm} \large\frac{100}{170}\hspace{0.2cm} \small \approx \hspace{0.2cm} 0.5882$<br>


- $\hspace{.3cm}P(\text{X=male} \mid \text{Y=blue}) \hspace{1.2cm}=\hspace{0.5cm}\large \frac{P(\text{X=male} \hspace{0.2cm}\cap\hspace{0.2cm} \text{Y=blue})}{P(\text{Y=blue})}       \hspace{0.9cm} \small= \hspace{0.5cm} \large \frac{20}{30}\hspace{0.5cm} \small\approx \hspace{0.2cm} 0.666$<br>


- $\hspace{.3cm}P(\text{X=female} \mid \text{Y=blue}) \hspace{0.9cm}=\hspace{0.5cm}\large \frac{P(\text{X=female} \hspace{0.2cm}\cap\hspace{0.2cm} \text{Y=blue})}{P(\text{Y=blue})}       \hspace{0.5cm} \small= \hspace{0.5cm} \large \frac{10}{30}\hspace{0.5cm} \small \approx \hspace{0.2cm} 0.333$<br>

**Compute different conditional probabilities of a randomly selected bird having brown/blue eyes given that it is  male/female:**<br><br>

- $\hspace{.3cm}P(\text{Y=brown} \mid \text{X=male}) \hspace{0.9cm}=\hspace{0.5cm}\large \frac{P(\text{Y=brown} \hspace{0.2cm}\cap\hspace{0.2cm} \text{X=male})}{P(\text{X=male})}       \hspace{0.5cm} \small= \hspace{0.5cm} \large \frac{70}{90}\hspace{0.8cm} \small \approx \hspace{0.2cm} 0.778$<br>


- $\hspace{.3cm}P(\text{Y=blue} \mid \text{X=male}) \hspace{1.3cm}=\hspace{0.5cm}\large \frac{P(\text{Y=blue} \hspace{0.2cm}\cap\hspace{0.2cm} \text{X=male})}{P(\text{X=male})}       \hspace{0.8cm} \small= \hspace{0.5cm} \large\frac{20}{90}\hspace{0.8cm} \small \approx \hspace{0.2cm} 0.222$<br>


- $\hspace{.3cm}P(\text{Y=brown} \mid \text{X=female}) \hspace{0.7cm}=\hspace{0.5cm}\large \frac{P(\text{Y=brown} \hspace{0.2cm}\cap\hspace{0.2cm} \text{X=female})}{P(\text{X=female})}       \hspace{0.3cm} \small= \hspace{0.5cm} \large \frac{100}{110}\hspace{0.5cm} \small\approx \hspace{0.2cm} 0.909$<br>


- $\hspace{.3cm}P(\text{Y=blue} \mid \text{X=female}) \hspace{1.1cm}=\hspace{0.5cm}\large \frac{P(\text{Y=blue} \hspace{0.2cm}\cap\hspace{0.2cm} \text{X=female})}{P(\text{X=female})}       \hspace{0.5cm} \small= \hspace{0.5cm} \large \frac{10}{110}\hspace{0.5cm} \small \approx \hspace{0.2cm} 0.0909$<br>

<h3 align="center"><div class="alert alert-success" style="margin: 20px">Conditional Probability is the probability of an event A occurring given that another event B has already occurred.</h3>
   
<font face = "Courier New" color=magenta size=5>
\begin{equation}
    P(A \mid B) = \frac{P(A \cap B)}{P(B)}
\end{equation}
</font>

>- The formula to calculate the conditional probability is equal to the joint probability of the two events divided by the marginal probability of the event that has already occurred

**Example 5:** Consider the Titanic Dataset with just two input features pclass and gender along with output column survived
- Given that a person belong to a specific class, determine the conditional probability of his survival
- Given the gender of a person, determine the conditional probability of his survival

**`Load the Dataset:`**

In [1]:
import numpy as np
import pandas as pd
df = pd.read_csv('datasets/titanic2.csv')
df

,survived,pclass,sex,age,sibsp,parch,fare,embarked,class,who,adult_male,deck,embark_town,alive,alone
0,0,3,male,22.0,1,0,7.2500,S,Third,man,True,NaN,Southampton,no,False
1,1,1,female,38.0,1,0,71.2833,C,First,woman,False,C,Cherbourg,yes,False
2,1,3,female,26.0,0,0,7.9250,S,Third,woman,False,NaN,Southampton,yes,True
3,1,1,female,35.0,1,0,53.1000,S,First,woman,False,C,Southampton,yes,False
4,0,3,male,35.0,0,0,8.0500,S,Third,man,True,NaN,Southampton,no,True
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
886,0,2,male,27.0,0,0,13.0000,S,Second,man,True,NaN,Southampton,no,True
887,1,1,female,19.0,0,0,30.0000,S,First,woman,False,B,Southampton,yes,True
888,0,3,female,NaN,1,2,23.4500,S,Third,woman,False,NaN,Southampton,no,False
889,1,1,male,26.0,0,0,30.0000,C,First,man,True,C,Cherbourg,yes,True


In [2]:
df1 = pd.concat([df['pclass'], df['sex'], df['survived']], axis=1)
df1

,pclass,sex,survived
0,3,male,0
1,1,female,1
2,3,female,1
3,1,female,1
4,3,male,0
...,...,...,...
886,2,male,0
887,1,female,1
888,3,female,0
889,1,male,1


**`Conditional Probabilities of survival given the pclass info of passenger:`**
>- Let X be a random variable associated with the `pclass` of a Titanic passenger and can take the values {1,2,3}
>- Let Y be a random variable associated with the `survived` status of a Titanic passenger and can take the values {0, 1} 

In [3]:
pd.crosstab(index=df['pclass'], columns=df['survived'], margins=True)

survived,0,1,All
pclass,,,
1,80,136,216
2,97,87,184
3,372,119,491
All,549,342,891


<img align="center" width="500"  src="images/nb-ex32.png"  >

- Find the conditional probability of a person `not surviving` given that he/she is travelling in `pclass 1`:
$$P(\text{survived}=0 \mid \text{pclass}=1) \hspace{0.5cm}=\hspace{0.5cm} \frac{P(\text{survived}=0 \hspace{0.2cm}\cap\hspace{0.2cm} \text{pclass}=1)}{P(\text{pclass}=1)}\hspace{0.5cm} = \hspace{0.5cm} \frac{80}{216}\hspace{0.5cm}= \hspace{0.5cm} 0.37037$$<br>

- Find the conditional probability of a person `surviving` given that he/she is travelling in `pclass 1`:
$$P(\text{survived}=1 \mid \text{pclass}=1) \hspace{0.5cm}=\hspace{0.5cm} \frac{P(\text{survived}=1 \hspace{0.2cm}\cap\hspace{0.2cm} \text{pclass}=1)}{P(\text{pclass}=1)}\hspace{0.5cm} = \hspace{0.5cm} \frac{136}{216}\hspace{0.5cm}= \hspace{0.5cm} 0.62963$$<br>

In [4]:
# by setting the normalize='index', you will get the conditional probabilities given pclass
pd.crosstab(index=df['pclass'], columns=df['survived'], normalize='index')

survived,0,1
pclass,,
1,0.370370,0.629630
2,0.527174,0.472826
3,0.757637,0.242363


> The above table shows the conditional probabilities
>- 0.37037 is the probability of a person `not surviving` given that he/she belong to pclass 1
>- 0.62963 is the probability of a person `surviving`, given that he/she is traveling in pclass 1. 
>- The sum of the two probabilities in that row is 1.
>- Similarly, you can check the rest of the probabilities as well.

**`Conditional Probabilities of survival given the gender info of passenger:`**
>- Let X be a random variable associated with the `gender` of a Titanic passenger and can take the values {female, male}
>- Let Y be a random variable associated with the `survived` status of a Titanic passenger and can take the values {0, 1} 

In [5]:
pd.crosstab(index=df['sex'], columns=df['survived'], margins=True)

survived,0,1,All
sex,,,
female,81,233,314
male,468,109,577
All,549,342,891


<img align="center" width="500"  src="images/nb-ex33.png"  >

- Find the conditional probability of a person `not surviving` given that she is `female`:
$$P(\text{survived}=0 \mid \text{sex}=\text{female}) \hspace{0.5cm}=\hspace{0.5cm} \frac{P(\text{survived}=0 \hspace{0.2cm}\cap\hspace{0.2cm} \text{sex}=\text{female})}{P(\text{sex}=\text{female})}\hspace{0.5cm} = \hspace{0.5cm} \frac{81}{314}\hspace{0.5cm}= \hspace{0.5cm} 0.257962$$<br>

- Find the conditional probability of a person `surviving` given that she is `female`:
$$P(\text{survived}=1 \mid \text{sex}=\text{female}) \hspace{0.5cm}=\hspace{0.5cm} \frac{P(\text{survived}=1 \hspace{0.2cm}\cap\hspace{0.2cm} \text{sex}=\text{female})}{P(\text{sex}=\text{female})}\hspace{0.5cm} = \hspace{0.5cm} \frac{233}{314}\hspace{0.5cm}= \hspace{0.5cm} 0.74203822$$<br>


In [6]:
# by setting the normalize='index', you will get the conditional probabilities given sex
pd.crosstab(index=df['sex'], columns=df['survived'], normalize='index')

survived,0,1
sex,,
female,0.257962,0.742038
male,0.811092,0.188908


> The above table shows the conditional probabilities
>- 0.257962 is the probability of a person `not surviving` given that she is female
>- 0.742038 is the probability of a person `surviving` given that she is female
>- The sum of the two probabilities in that row is 1.
>- Similarly, you can check the rest of the probabilities as well.